File: 14 - EdgeModules Extending RAD Server/C++/EdgeModules Example/EdgeModuleU.cpp
//---------------------------------------------------------------------------
#pragma hdrstop

#include "EdgeModuleU.h"
#include <memory>
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma classgroup "FMX.Controls.TControl"
#pragma resource "*.dfm"
//---------------------------------------------------------------------------

// Initialize global variable
EdgeModuleU::TFormValues EdgeModuleU::FormValues = { false, L"", 0.0 };

__fastcall TProcessorResource1::TProcessorResource1(TComponent* Owner)
	: TDataModule(Owner)
{
}

void TProcessorResource1::GetCalculate(TEndpointContext* AContext,
	TEndpointRequest* ARequest, TEndpointResponse* AResponse)
{
    UnicodeString LParam1, LParam2;
    TJSONObject* LResponse;

	// Read the input params
	if (!ARequest->Params->TryGetValue("param1", LParam1) ||
		!ARequest->Params->TryGetValue("param2", LParam2))
	{
		AResponse->RaiseBadRequest("Bad request", "Missing data");
    }

    LResponse = new TJSONObject();
    try
    {
        // Calculate the result
        double calculation = StrToFloat(LParam1) + StrToFloat(LParam2);

        // Add your processing logic here
        LResponse->AddPair("status", "processed");
        LResponse->AddPair("timestamp", DateToISO8601(Now()));
        LResponse->AddPair("result", calculation);

        // Send the response
		AResponse->Body->SetValue(LResponse, false);
    }
    __finally
    {
        LResponse->Free();
    }
}

void TProcessorResource1::GetFormValues(TEndpointContext* AContext,
    TEndpointRequest* ARequest, TEndpointResponse* AResponse)
{
	AResponse->Body->SetValue(EdgeModuleU::FormValues.ToJSON(), true);
}

// TFormValues implementation
TJSONObject* EdgeModuleU::TFormValues::ToJSON()
{
    TJSONObject* Result = new TJSONObject();
    Result->AddPair("CheckBox", this->CheckBox);
    Result->AddPair("Edit", this->Edit);
    Result->AddPair("TrackBar", this->TrackBar);
    return Result;
}

static void Register()
{
    std::unique_ptr<TEMSResourceAttributes> attributes(new TEMSResourceAttributes());
	attributes->ResourceName = "processor";
	attributes->ResourceSuffix["GetCalculate"] = "calculate";
	attributes->ResourceSuffix["GetFormValues"] = "formValues";
	RegisterResource(__typeinfo(TProcessorResource1), attributes.release());
}

#pragma startup Register 64

File: 14 - EdgeModules Extending RAD Server/Delphi/EdgeModules Example/EdgeModuleU.pas
unit EdgeModuleU;

// EMS Resource Module

interface

uses
  System.SysUtils, System.Classes, System.JSON,
  EMS.Services, EMS.ResourceAPI, EMS.ResourceTypes,
  System.DateUtils;

type
  TFormValues = record
    CheckBox: boolean;
    Edit: string;
    TrackBar: double;
    function ToJSON: TJSONObject;
  end;

  [ResourceName('processor')]
  TProcessorResource1 = class(TDataModule)
  published
    [EndPointRequestSummary('Processor', 'GetCalculate', 'Retrieves item with specified ID', 'application/json', '')]
    [EndPointRequestParameter(TAPIDocParameter.TParameterIn.Path, 'param1', 'Parameter nº 1', true, TAPIDoc.TPrimitiveType.spNumber,
      TAPIDoc.TPrimitiveFormat.None, TAPIDoc.TPrimitiveType.spNumber, '', '')]
    [EndPointRequestParameter(TAPIDocParameter.TParameterIn.Path, 'param2', 'Parameter nº 2', true, TAPIDoc.TPrimitiveType.spNumber,
      TAPIDoc.TPrimitiveFormat.None, TAPIDoc.TPrimitiveType.spNumber, '', '')]
    [EndPointResponseDetails(200, 'Ok', TAPIDoc.TPrimitiveType.spObject, TAPIDoc.TPrimitiveFormat.None, '', '')]
    [ResourceSuffix('calculate')]
    procedure GetCalculate(const AContext: TEndpointContext;
                      const ARequest: TEndpointRequest;
                      const AResponse: TEndpointResponse);
    [ResourceSuffix('formValues')]
    procedure GetFormValues(const AContext: TEndpointContext;
                      const ARequest: TEndpointRequest;
                      const AResponse: TEndpointResponse);
  end;

var FormValues: TFormValues;

implementation

{%CLASSGROUP 'FMX.Controls.TControl'}

{$R *.dfm}

procedure TProcessorResource1.GetCalculate(const AContext: TEndpointContext;
  const ARequest: TEndpointRequest; const AResponse: TEndpointResponse);
var
  LParam1, LParam2: string;
  LResponse: TJSONObject;
begin
  // Read the input params
    if not(ARequest.Params.TryGetValue('param1', LParam1)) or
       not(ARequest.Params.TryGetValue('param2', LParam2)) then
    AResponse.RaiseBadRequest('Bad request', 'Missing data');
    LResponse := TJSONObject.Create;
    try
      // Calculate the result
      var calculation := LParam1.ToDouble + LParam2.ToDouble;
      // Add your processing logic here
      LResponse.AddPair('status', 'processed');
      LResponse.AddPair('timestamp', DateToISO8601(Now));
      LResponse.AddPair('result', calculation);
      // Send the response
      AResponse.Body.SetValue(LResponse, False);
    finally
      LResponse.Free;
    end;
end;

procedure TProcessorResource1.GetFormValues(const AContext: TEndpointContext;
  const ARequest: TEndpointRequest; const AResponse: TEndpointResponse);
begin
  AResponse.Body.SetValue(FormValues.ToJSON, True);
end;

{ TFormValues }

function TFormValues.ToJSON: TJSONObject;
begin
  Result := TJSONObject.Create;
  Result.AddPair('CheckBox', self.CheckBox);
  Result.AddPair('Edit', self.Edit);
  Result.AddPair('TrackBar', self.TrackBar);
end;

procedure Register;
begin
  RegisterResource(TypeInfo(TProcessorResource1));
end;

initialization
  Register;
end.


